server/tserver/src/main/java/org/apache/accumulo/tserver/TabletServer.java

```diff
@@ -394,6 +398,20 @@ private static long jitter() {
         * TabletServer.TIME_BETWEEN_LOCATOR_CACHE_CLEARS);
   }
 
+  public Semaphore getSemaphore() {
+    int writeThreads =
+        getServerConfig().getConfiguration().getCount(Property.TSERV_MAX_WRITETHREADS);
+    if (writeThreads == 0)
+      writeThreads = Integer.MAX_VALUE;
+    if (sem == null || maxThreadPermits != writeThreads) {
```

Comment:
sem and maxThreadPermits are read outside sync, should probably be volatile. Also the checks should be done again inside the sync block, because multuple threads could execute the if and get a true value and then block on the sync block.